[TFLite] Fully quantize the uint8 concatenation kernel to make it pure integer

Tessil opened the pull request:
Hi,

This PR fixes the issue raised by #37544 and fully quantizes the uint8 concatenation kernel in TFLite and TFLite micro so that it only uses integer operations..

It pre-calculates the scaling multiplier and shift with QuantizeMultiplier and then uses MultiplyByQuantizedMultiplier instead of a floating-point multiplication.

Thibaut

Tessil committed:
[TFLite] Fully quantize the uint8 concatenation kernel to make it pure integer

suharshs commented:
Since there are models running with the existing rescaling operation, the change could theoretically cause a change in accuracy (or even just a change in one input's prediction). Therefore, if we add this change we should probably add an attribute to the operation that signals a version change (similar to what @wwwind  has done here https://github.com/tensorflow/tensorflow/pull/35985) , and set the corresponding version in the converter.

gbaned commented:
@Tessil Can you please check @suharshs comments and keep us posted. Thanks!

Tessil commented:
Hi,

Thank you for your answer. Changing from a floating-point scaling to a fixed-point scaling may effectively cause a small change in accuracy though it should be quite rare. I am currently creating a new version of the concatenation operator with an additional `fixed_point_scaling:bool = false` field in the flatbuffers to avoid any accuracy change with older tflite files. I will update the PR when everything is ready.

Tessil committed:
Add a fixed_point_scaling parameter with false as default value to the TFLite concatenation operator to keep backward compatibiltiy. This create the version 4 of the operator.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil commented:
Hi,

I created a new version for the concatenation operator. I used 4 as new version number to avoid a future conflict with PR #38869 which creates a version 3 of the same operator. This can be changed if necessary.

The new version has an extra `fixed_point_scaling:bool` field in the TFLite flatbuffer. It defaults to 'false' when importing an old '.tflite' file without this field and to 'true' with new exports.

I followed the documentation regarding the [versioning of the operators](https://www.tensorflow.org/lite/guide/ops_version) though it doesn't seem totally complete. Let me know if there is something that I have forgotten regarding the versioning.

Tessil commented on tensorflow/compiler/mlir/lite/ir/tfl_ops.td:
```diff
@@ -644,7 +644,8 @@ def TFL_ConcatenationOp : TFL_Op<"concatenation",
     I32Attr:$axis,
-    TFL_AFAttr:$fused_activation_function
+    TFL_AFAttr:$fused_activation_function,
+    BoolAttr:$fixed_point_scaling
```
Is it necessary to add a `fixed_point_scaling` field for the MLIR converter? From what I understand, the `SameOperandsAndResultsScale` trait of `TFL_ConcatenationOp` guarantees that the scaling of the inputs and output will be the same . In this case the concatenation kernel for uint8 only does a memcpy regardless of the value of the `fixed_point_scaling` in the TFLite flatbuffer.

liufengdb replied:
right. the scales will be the same and the kernel will use memory copy.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Use BoolAttr instead of auto.

gbaned commented:
@Tessil  Can you please resolve conflicts? Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Adapt new test to the new fixed_point_scaling field.

Tessil commented:
Done, I also adapted the new concatv2I64Axis test that was added in the master branch.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

suharshs commented:
Feng, could you verify if the changes for the mlir converter are required. Or if because of the same scales trait, we can avoid needing to set this field in the mlir converter.

thanks!

Tessil committed:
Remove unnecessary fixed_point_scaling field in the MLIR concatenation operator.

Tessil commented:
Thank you for the confirmation regarding the `SameOperandsAndResultsScale` trait, I removed the `fixed_point_scaling` field of the MLIR `TFL_ConcatenationOp` to simplify the code. As the trait guarantees that the input and output scales will be the same, it doesn't matter if the `fixed_point_scaling` is set to true for the MLIR converter, the kernel will just do a `memcpy` in any case as it doesn't need any scaling.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil commented:
@suharshs I resolved a merge conflict that arised, the PR will need a re-approval.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

gbaned commented:
@Tessil Can you please resolve conflicts? Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil commented:
@gbaned Done, it will again need re-approval. Thanks.

liufengdb commented:
Looks like this is a very general optimization. I'm wondering whether we have a general design for this. 

For example, we can make two optional fields in the quantization parameters of the flatbuffer tensors. Then the fixed point quantization parameter there. Then we can update all the kernels to support this optional fields.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Fix merge, rename concatenation_params to schema_params.

Tessil commented:
I am not sure it is worth to generalize the `fixed_point_scaling` parameter. Looking at some of the other layers the usage of a floating-point scaling seems to be an exception here, other layers tend to use a fixed-point scaling. 

I resolved a conflict that arised again, the PR will need re-approval if we don't plan to generalise the fixed-point scaling parameter. Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

gbaned commented:
@Tessil Can you please resolve conflicts? Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Don't use fixed point scaling for Concatenation legacy op.

Tessil commented:
Hi,

@gbaned Thanks, I fixed the the merge conflict.

I also changed the `tflite::ops::micro::concatenation::Prepare` switch case introduced by 6a22ad8f67f25a5a321ff6306ccbac4a3e33ee19 a bit. Setting-up the scalings for int8 kernels is not necessary as `EvalUnquantized` is called with int8 anyway.

@liufengdb, @suharshs Could you take a look when you have a bit of time, the merge conflicts start to accumulate.

Thanks,
Thibaut

Tessil committed:
Add new Concatenation op runtime version.

gbaned commented:
@Tessil Can you please resolve conflicts? Thanks!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/quantized-uint8-concat

Tessil commented:
Hi,

@gbaned, @suharshs I fixed the merge conflicts, the PR will need re-approval. Thanks.

Tessil commented:
@gbaned It seems there is an error with copybara "An error happened while migrating the change", is there anything I can do? Thanks.

gbaned commented:
> @gbaned It seems there is an error with copybara "An error happened while migrating the change", is there anything I can do? Thanks.

@Tessil Nothing to do at this moment from your side. We will come back if anything required. Thank you!

gbaned commented:
@Tessil  Can you please resolve conflicts? Thanks!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/quantized-uint8-concat

Tessil commented:
@gbaned I fixed the conflicts. It seems though that once the PR is approved and all the tests are successful, the PR is not merged automatically even after staying for one week in the 'ready to pull' state. Could you eventually check if there is a problem internally regarding the PR? Thanks!

gbaned commented:
> @gbaned I fixed the conflicts. It seems though that once the PR is approved and all the tests are successful, the PR is not merged automatically even after staying for one week in the 'ready to pull' state. Could you eventually check if there is a problem internally regarding the PR? Thanks!

@Tessil  This is waiting for an internal approval.  We will let you know if anything required from you. Thank you!

gbaned commented:
@Tessil  Can you please resolve conflicts? Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil commented:
@gbaned @rthadur I fixed the merge conflict, the PR will need re-approval. Thanks.

gbaned commented:
@jianlijianli, @daverim Can you please have a look at this PR? Thank you!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil commented:
Fixed a merge conflict that arisen. The PR will need re-approval, thanks.

talumbau commented:
The TFLite micro build is failing on the `using tflite::testing::F2Q` statement. I can't find the definition of this symbol, nor the name `tflite::testing::F2QS` used elsewhere in Micro. Adding a Micro team member to take a look at the CI failure and advise.

talumbau commented:
@advaitjain can you take a look at the Micro build failure and advise? Thanks very much!

Tessil commented:
Thank you very much for looking into it. It seems `F2Q` was removed by https://github.com/tensorflow/tensorflow/commit/d39ff659206f7a08583608f7971abd747f5f6709 26 days ago and I missed it during the merge conflict resolution. I will check to update the PR tomorrow.

advaitjain commented:
There might be a few additional micro-specific things to fix.

Please use the following command to reproduce locally:
```
make -f tensorflow/lite/micro/tools/make/Makefile test_kernel_concatenation_test
```

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Adapt TFL and TFLµ concatenation tests to the changes from the master branch

Tessil commented:
@talumbau I adapted the TFL and TFLµ tests to the changes from the master branch and tested them locally. Let me know if there is any problem. Thanks!

Tessil committed:
Fix missing initializer member warning in TFLµ concatenation test

Tessil commented:
@talumbau I fixed a problem in the TFµ test, the PR will need re-approval, sorry. 

I used `bazel //tensorflow/lite/kernels:concatenation_test //tensorflow/lite/micro/kernels:concatenation_test` to test locally instead of `make -f tensorflow/lite/micro/tools/make/Makefile test_kernel_concatenation_test` and it didn't show the warning.

advaitjain reviewed (approved):
looks good for the changes for Micro.

talumbau commented:
The Linux GPU and TFLite Micro CI build failures seem spurious. Is that what is blocking the copybara import? Will have to ask around to get an answer.

talumbau commented:
I approved the copybara import and can now see the internal change up for review.

rthadur commented:
@Tessil can you please check below error 

```
tensorflow/lite/micro/kernels/concatenation.cc:227:46: error: implicit conversion increases floating-point precision: 'float' to 'double' 
          QuantizeMultiplier(input_scales[i] * inverse_output_scale,
```

Tessil committed:
Avoid implicit conversion from float to double in TFLµ concatenation prepare

Tessil commented:
@rthadur Thank you, I fixed the error.

talumbau commented:
Thanks for the quick fix! There is one last question that came up in the review. It appears that `fixed_point_scaling` is never set to `true` in the MLIR conversion path. The default setting is `true` for a toco-based conversion path, which is now deprecated. Can you explain your intended usage of this flag?

Overall, it appears that, if one is still using a toco-based path to convert, one would get this flag set to true as default, but if one is using the MLIR-based conversion path, one would get this flag as false. If that is correct, that is probably not the best path to go down. In other words, it seems not desirable to add new behavior to a deprecated path. Please correct me if I have misunderstood what's happening here. Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Set concatenation fixed_point_scaling parameter to true by default for new MLIR-converted models

Tessil commented:
I initially added a `fixed_point_scaling` parameter to the MLIR `TFL_ConcatenationOp` which defaulted to true (see https://github.com/tensorflow/tensorflow/pull/38704#discussion_r418516214) but it seems it wasn't necessary as the new MLIR converter forces the scaling of the input and output to be the same. As far as I know it isn't possible with the MLIR converter to create a model with an uint8 concatenation layer that needs rescaling.

This PR is quite old and was initially created when the TOCO converter was still kind of used. As the TOCO converter is now completely deprecated and if what I said above is correct, we can probably abandon this PR as in the end I don't think it's really useful any more. I'll check in intern if we can close this.

akarmi commented:
I believe the main reason for this PR is captured in the first line of its description - to fix the reference implementation so that it does not contain floating point operations. The `fixed_point_scaling` flag was introduced to distinguish between the old and new version of the operator. Changing the behaviour of the old converter and not the MLIR-based path is the result of the [this disucssion](https://github.com/tensorflow/tensorflow/pull/38704#issuecomment-637745934), and seems to be an unfortunate artefact of the fact that two converters treat relation between input and output scales differently (following the change highlighted [here](https://github.com/tensorflow/tensorflow/issues/37544#issuecomment-658580281), if I follow things correctly).

@dominicsymes, what is the best way forward here from your point of view?

Tessil commented:
Oh yes, effectively. @liufengdb added an extra `extraClassDeclaration` with the https://github.com/tensorflow/tensorflow/commit/ac9969a2762d4068edef7e451b53256d2adb2692 commit later on. The comment above (https://github.com/tensorflow/tensorflow/pull/38704#discussion_r418516214) is thus not valid anymore and we need to re-add the `fixed_point_scaling` parameter in the `TFL_ConcatenationOp`. I will check to update the PR today.

Tessil commented:
@talumbau I updated the PR, the `fixed_point_scaling` is now set to true for both new TOCO and MLIR converted models as it was done before. Please let me know if there is any problem, thanks.

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

talumbau commented:
Thanks for the MLIR changes. Looks good to me!

Tessil committed:
Apply clang-format to TFLite micro concatenation tests

gbaned commented:
@Tessil  Can you please resolve conflicts? Thanks!

Tessil committed:
Merge branch 'master' into toupstream/quantized-uint8-concat

Tessil commented:
@gbaned Resolved, thanks!

gbaned commented:
@Tessil Can you please resolve conflicts? Thanks!

Tessil commented:
With uint8 support being removed from TFLite Micro and deprecated in TFLite since some time, the PR is not much more relevant any more and it's probably not worth to spend more time on it. I will close the PR.
